controllers/csi/daemonset.go

```diff
@@ -0,0 +1,326 @@
+)
+
+type ReconcileCSI struct {
+	client   client.Client
+	scheme   *runtime.Scheme
+	logger   logr.Logger
+	instance *v1alpha1.DynaKube
+}
+
+func NewCSIReconciler(client client.Client, scheme *runtime.Scheme, logger logr.Logger, instance *v1alpha1.DynaKube) *ReconcileCSI {
+	return &ReconcileCSI{
+		client:   client,
+		scheme:   scheme,
+		logger:   logger,
+		instance: instance,
+	}
+}
+
+func (r *ReconcileCSI) Reconcile() (bool, error) {
+	r.logger.Info("Reconciling CSI driver")
+
+	ds, err := buildDesiredCSIDaemonSet()
+	if err != nil {
+		return false, errors.WithStack(err)
+	}
+
+	if err := controllerutil.SetControllerReference(r.instance, ds, r.scheme); err != nil {
+		return false, errors.WithStack(err)
+	}
+
+	upd, err := utils.CreateOrUpdateDaemonSet(r.client, r.logger, ds)
+	if upd || err != nil {
+		return upd, errors.WithStack(err)
+	}
+
+	return false, nil
+}
+
+func buildDesiredCSIDaemonSet() (*appsv1.DaemonSet, error) {
+	metadata := prepareMetadata()
+	driver := prepareDriverSpec()
+	registrar := prepareRegistrarSpec()
+	livenessprobe := preparelivenessProbeSpec()
+	vol := prepareVolumes()
+
+	ds := &appsv1.DaemonSet{
+		ObjectMeta: metadata,
+		Spec: appsv1.DaemonSetSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"internal.oneagent.dynatrace.com/component": "csi-driver",
+					"internal.oneagent.dynatrace.com/app":       "csi-driver",
+				},
+			},
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{
+						"kubectl.kubernetes.io/default-logs-container": "driver",
+					},
+					Labels: map[string]string{
+						"internal.oneagent.dynatrace.com/component": "csi-driver",
+						"internal.oneagent.dynatrace.com/app":       "csi-driver",
+					},
+				},
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{
+						driver,
+						registrar,
+						livenessprobe,
+					},
+					ServiceAccountName: DaemonSetName,
+					Volumes:            vol,
+				},
+			},
+		},
+	}
+
+	dsHash, err := generateDaemonSetHash(ds)
+	if err != nil {
+		return nil, err
+	}
+	ds.Annotations[activegate.AnnotationTemplateHash] = dsHash
+
+	return ds, nil
+}
+
+func generateDaemonSetHash(ds *appsv1.DaemonSet) (string, error) {
+	data, err := json.Marshal(ds)
+	if err != nil {
+		return "", err
+	}
+
+	hasher := fnv.New32()
+	_, err = hasher.Write(data)
+	if err != nil {
+		return "", err
+	}
+
+	return strconv.FormatUint(uint64(hasher.Sum32()), 10), nil
+}
+
+func prepareMetadata() metav1.ObjectMeta {
+	return metav1.ObjectMeta{
+		Name:      DaemonSetName,
+		Namespace: "dynatrace",
+		Labels: map[string]string{
+			"dynatrace.com/operator": "dynatrace",
+		},
+		Annotations: map[string]string{},
+	}
+}
+
+func prepareDriverSpec() v1.Container {
+	privileged := true
+	userID := int64(0)
+	bidirectional := v1.MountPropagationBidirectional
+
+	return v1.Container{
+		Name:    "driver",
+		Image:   "quay.io/dynatrace/dynatrace-operator:snapshot",
```

Comment:
This should be the same as the operator image or configurable.

Reply:
Changed to query the operator pod and reusing the image from there